File: infrastructures/ddd/python/src/infrastructure/queues/infra/i_event_publisher.py
"""
Interface for publishing domain events to a message queue.
Used for asynchronous communication between subdomains.
CRITICAL: Domain events via queue are the ONLY way to communicate between subdomains.
HTTP calls between subdomains are FORBIDDEN.
"""

from abc import ABC, abstractmethod
from typing import TypeVar


TEvent = TypeVar('TEvent')


class IEventPublisher(ABC):
    """
    Interface for publishing domain events.
    
    Topic naming convention: {subdomain}.{action}
    Examples: "example.created", "example.updated", "example.deleted"
    
    Used for:
    - Asynchronous subdomain-to-subdomain communication
    - Event-driven architecture
    - Domain event broadcasting
    """
    
    @abstractmethod
    async def publish_async(self, p_event: TEvent, p_topic: str) -> None:
        """
        Publishes a domain event to the specified topic.
        
        Args:
            p_event: The event to publish
            p_topic: The topic to publish to
        """
        pass

File: infrastructures/ddd/java/src/main/java/com/oddly/ddd/infrastructure/queues/infra/IEventPublisher.java
package com.oddly.ddd.infrastructure.queues.infra;

import java.util.concurrent.CompletableFuture;

/**
 * Interface for publishing domain events to a message queue.
 * Event publishers:
 * - Live in /infrastructure/queues/infra/
 * - Used for asynchronous communication between subdomains
 * - MUST be used for ALL subdomain-to-subdomain communication
 * - MUST NOT use HTTP calls between subdomains
 * 
 * Topic naming: {subdomain}.{action}
 * Example: example.created, example.updated
 */
public interface IEventPublisher {
    /**
     * Publishes an event to the specified topic.
     * 
     * @param <TEvent> The event type
     * @param p_event The event to publish
     * @param p_topic The topic to publish to
     */
    <TEvent> CompletableFuture<Void> publishAsync(TEvent p_event, String p_topic);
}
